libtenzir/src/catalog.cpp

```diff
@@ -86,17 +87,32 @@ void catalog_state::erase(const uuid& partition) {
 }
 
 caf::expected<catalog_lookup_result>
-catalog_state::lookup(const expression& expr) const {
+catalog_state::lookup(expression expr) const {
   auto start = stopwatch::now();
   auto total_candidates = catalog_lookup_result{};
   auto num_candidates = size_t{0};
-  auto pruned = prune(expr, unprunable_fields);
+  auto normalized = normalize_and_validate(expr);
+  if (not normalized) {
+    return caf::make_error(ec::invalid_argument,
+                           fmt::format("{} failed to normalize and validate "
+                                       "epxression {}: {}",
+                                       *self, expr, normalized.error()));
+  }
+  expr = std::move(*normalized);
+  expr = prune(expr, unprunable_fields);
+  if (expr == caf::none) {
+    expr = trivially_true_expression();
+  }
   for (const auto& [type, _] : synopses_per_type) {
-    auto resolved = resolve(taxonomies, pruned, type);
-    if (!resolved) {
-      return resolved.error();
+    auto resolved = resolve(taxonomies, expr, type);
+    if (not resolved) {
+      return caf::make_error(ec::invalid_argument,
+                             fmt::format("{} failed to resolve epxression {}: "
+                                         "{}",
+                                         *self, expr, resolved.error()));
```

Comment:
I noticed a few very peculiar things about this as part of my investigation into the failing integration test.

1. `resolve` should be called _before_ `prune`, as otherwise `prune` may be ineffective if the expression contains concepts.
2. `resolve` only resolves for the first type that it encounters. That is, if I import a `zeek.dns` before a `suricata.flow`, then `net.src.ip` resolves to `zeek.dns.id.orig_h` even for `suricata.flow` partitions.

(1) is a missed optimization, (2) is a hard bug. I think this only worked previously because always returned all candidate partitions here, and the evaluation logic works correctly.

@tobim can you take this over and then merge this, since I won't have time to take another look at this until the end of the week?

Reply:
I managed to find (2) myself, this was a shadowing mistake. (1) I fixed as well.